### PR TITLE
Add visual feedback when changing grid size

### PR DESCRIPTION
**Pull request by Karolis-Sakavicius**
Currently when we use a hotkey to change a grid size, there is no visual or any other feedback indicating which grid size was selected. This PR introduces a new setting, which allows users to see grid lines when they change the grid size.

---

This is not the greatest way of handling this - if one is moving through grid sizes faster than 1 sec/1 change, grid lines will overlap.

I originally intended to make it without a delay - when hotkey is pressed, grid lines are shown and when the keys are released, they are removed. But seems like GNOME extension APIs are very poorly documented, so I couldn't find a way to make it work with key press & key release events. :( If somebody could help me out with this, I'd really appreciate it and change it.

Also, I have only one monitor, so if somebody could test this with 2-monitor setup, it would be very helpful. :pray:

### CODE REVIEW

**Comment by scherepanov** [comment]
It used to work just fine. You activate gTile GUI, and it is coming up with grid according to current grid size. When you press space (default keyboard shortcut for changing grid) grid changed to next settings in a list. No delays.
It kinda break around a year ago, with some big change.
In theory, original functionality provides very clear visual feedback for users.
Yes, gnome-shell API is not very well documented.
Will check your PR.

**Comment by scherepanov** [comment]
Yes, feature is cool!

Few things:

1) Grid lines on a screen are shifted around 40 pixels to the right on my monitor. There is an empty vertical space adjacent to left monitor border. gTile do show preview of window placement on a grid, and this preview is off by same pixels, shifted same to the right, to grid lines.

2) What is not working - button with grid size, on GUI, is not set into active status. You used to be able to see what grid size are selected, button with grid size was activated. This is exactly what was broken some time ago. Can you look and fix it.

3) Please add few words to docs (README.md).

Apart from that, looks very nice!

**Comment by scherepanov** [comment]
Yes, I completely forget about using keyboard shortcut for changing grid size. It is very useful.

Please look on grid position - on my PC it is shifted by about 50 pixels to the right. You can see it cleanly, if you will activate gTile GUI, and then press "space" to change grid size.

Also, would be very nice if you can look on making buttons with grid sizes to look activated for current grid size, on gTile GUI.

**Comment by Karolis-Sakavicius** [comment]
@scherepanov here we go! :)

Please test things out - currently active grid button should be highlighted when summoning the UI, as well as when changing it through the UI or keyboard shortcut (`space` or global shortcut). Left offset bug should be fixed as well.

**Comment by scherepanov** [comment]
Works fine now!

With exception:
In multi-monitor configuration, grid lines are not shown on second monitor, but on primary. Primary monitor get both grids, for both monitors. That is true even if focus window is on second monitor.

Yes, quickly pressing keyboard shortcut creates not a very good picture of overlapping lines. I would not be able to help, that needs way too much digging into gnome-shell libraries.

Highlighting buttons with grid sizes works fine now!

Docs are fine.

Can I ask you to make one more change that would be outside of your scope. Preferences dialog looks very confusing, because it starts with too small window size. Can you change default size of preferences dialog, it will show all controls on basic tab. Should not be very hard.

We are almost ready to go merge. Last thing left - on multi-monitor all grid lines show only onprimary monitor.

File: app.ts [hunk 1]
```diff
@@ -604,6 +612,43 @@ class App {
...
+                    const newGridWidget = new St.BoxLayout({ style_class: `${theme}__grid_lines_preview` });
+                    gridTiles.push(newGridWidget);
+
+                    Main.uiGroup.add_actor(newGridWidget);
+
+                    newGridWidget.width = tileWidth;
+                    newGridWidget.height = tileHeight;
+                    newGridWidget.x = tileWidth * (i - 1);
+                    newGridWidget.y = topOffset + tileHeight * (u - 1);
+                    newGridWidget.monitor = activeMonitors()[monitorIdx];
```
Comment (Karolis-Sakavicius) [inline app.ts:639]: I added this but I have no way to test this out, as I have only one monitor. Maybe you @scherepanov could take a look and guide me to the right direction (or add an edit that would work)?

**Comment by scherepanov** [comment]
Just checked. Your last commit still have same problem - grid lines are being shown only on primary monitor.
Would be helpful if you will add logging, Something like printing properties of monitor where grid lines about to be displayed. Log properties like screen offset X and Y, screen width and height, value of monitorIdx.
I will be able to look on it tomorrow.
Thank you for working on it!

**Comment by scherepanov** [comment]
Tweaked with code. Everything looks fine, but all grid lines are only on primary monitor.
Is it correct that actor cannot change monitor after it has been created?
I tried to set monitor in constructor of St.BoxLayout. Still does not work.
Weird.

**Comment by Karolis-Sakavicius** [comment]
Managed to get my hands on a crappy second monitor. :grin: 

Turns out it doesn't show lines on primary monitor but rather on the leftmost monitor configured in X. And the way it works is not by defining a monitor where we want to draw an actor but rather we need to calculate the coordinates of the "total screen estate". E.g. if we have a dual monitor setup and want to show something on the second monitor, we need to do `firstMonitor.x + <our x coords>` (assuming that monitors are set up in a row, where first one is on the left and other one is on the right).

This is how it works with the dialog box as well - https://github.com/gTile/gTile/blob/master/app.ts#L535
So I did this: https://github.com/gTile/gTile/pull/262/files#diff-84dd6a83459413d5bb638cd885f20ef0dbebf2ebf10d0f35d1c9ca424b0ca06aR639
And it seems to work (hopefully not only on my machine).

Weird behavior indeed. :man_shrugging:

**Comment by scherepanov** [comment]
Now, with recent changes, I do not see grid lines from second monitor at all.
When window is on primary monitor (which happens to be on left, i.e. lower X), it does show grid lines on primary monitor only.
When focus window is on second monitor, grid lines are not showing on any monitors.

**Comment by Karolis-Sakavicius** [comment]
Hmm... I can't seem to replicate this. The lines are drawn no matter where the focus window is, with global shortcut or when pressing `space` while having the widget open (different gnome shell versions? Some dual monitor weirdness in X itself?).

I'll add some logging, so we can investigate further. I'm pretty much out of ideas at this point.

When I'm changing the grid size to 3:3, I get this: 
```
gTile Starting to draw grid lines for monitor {"index":0,"x":0,"y":0,"width":3440,"height":1440,"geometry_scale":1}
gTile Grid line of size 1146.6666666666667:460 is drawn at 0:27 (monitor offset 0:0)
gTile Grid line of size 1146.6666666666667:460 is drawn at 0:487 (monitor offset 0:0)
gTile Grid line of size 1146.6666666666667:460 is drawn at 0:947 (monitor offset 0:0)
gTile Grid line of size 1146.6666666666667:460 is drawn at 1146.6666666666667:27 (monitor offset 0:0)
gTile Grid line of size 1146.6666666666667:460 is drawn at 1146.6666666666667:487 (monitor offset 0:0)
gTile Grid line of size 1146.6666666666667:460 is drawn at 1146.6666666666667:947 (monitor offset 0:0)
gTile Grid line of size 1146.6666666666667:460 is drawn at 2293.3333333333335:27 (monitor offset 0:0)
gTile Grid line of size 1146.6666666666667:460 is drawn at 2293.3333333333335:487 (monitor offset 0:0)
gTile Grid line of size 1146.6666666666667:460 is drawn at 2293.3333333333335:947 (monitor offset 0:0)
gTile Starting to draw grid lines for monitor {"index":1,"x":3440,"y":0,"width":1440,"height":900,"geometry_scale":1}
gTile Grid line of size 480:300 is drawn at 3440:0 (monitor offset 3440:0)
gTile Grid line of size 480:300 is drawn at 3440:300 (monitor offset 3440:0)
gTile Grid line of size 480:300 is drawn at 3440:600 (monitor offset 3440:0)
gTile Grid line of size 480:300 is drawn at 3920:0 (monitor offset 3440:0)
gTile Grid line of size 480:300 is drawn at 3920:300 (monitor offset 3440:0)
gTile Grid line of size 480:300 is drawn at 3920:600 (monitor offset 3440:0)
gTile Grid line of size 480:300 is drawn at 4400:0 (monitor offset 3440:0)
gTile Grid line of size 480:300 is drawn at 4400:300 (monitor offset 3440:0)
gTile Grid line of size 480:300 is drawn at 4400:600 (monitor offset 3440:0)
gTile Removing grid lines...
```

What do you get?

**Comment by scherepanov** [comment]
This is when window is on second monitor. No grid lines shown at all:

Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Starting to draw grid lines for monitor {"index":0,"x":3840,"y":1098,"width":1680,"height":1050,"geometry_scale":1}
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 3840:2196 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 3840:2546 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 3840:2896 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 4400:2196 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 4400:2546 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 4400:2896 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 4960:2196 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 4960:2546 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 560:350 is drawn at 4960:2896 (monitor offset 3840:1098)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Starting to draw grid lines for monitor {"index":1,"x":0,"y":0,"width":3840,"height":2160,"geometry_scale":1}
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 0:48 (monitor offset 0:0)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 0:752 (monitor offset 0:0)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 0:1456 (monitor offset 0:0)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 1280:48 (monitor offset 0:0)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 1280:752 (monitor offset 0:0)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 1280:1456 (monitor offset 0:0)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 2560:48 (monitor offset 0:0)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 2560:752 (monitor offset 0:0)
Jan 02 20:02:51 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1280:704 is drawn at 2560:1456 (monitor offset 0:0)

This is when window is on primary monitor on left. Grid lines shows only on primary monitor (on left):

Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Started drawing grid lines...
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Starting to draw grid lines for monitor {"index":0,"x":3840,"y":1098,"width":1680,"height":1050,"geometry_scale":1}
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 840:525 is drawn at 3840:2196 (monitor offset 3840:1098)
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 840:525 is drawn at 3840:2721 (monitor offset 3840:1098)
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 840:525 is drawn at 4680:2196 (monitor offset 3840:1098)
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 840:525 is drawn at 4680:2721 (monitor offset 3840:1098)
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Starting to draw grid lines for monitor {"index":1,"x":0,"y":0,"width":3840,"height":2160,"geometry_scale":1}
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1920:1056 is drawn at 0:48 (monitor offset 0:0)
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1920:1056 is drawn at 0:1104 (monitor offset 0:0)
Jan 02 20:05:57 chi-fed26-scherepanov gnome-shell[240883]: gTile Grid line of size 1920:1056 is drawn at 1920:48 (monitor offset 0:0)

**Comment by scherepanov** [comment]
Wait. In both cases, focus window either on primary monitor on left, or secondary monitor on right, I can see grid lines. Only on primary monitor.
Sorry for confusion.

**Comment by scherepanov** [comment]
Well.
I am running remote desktop for linux. I do not have a physical access to linux box.
Remote desktop is set to use all available displays.
I do get remote desktop on all monitors, and it does work like I have two monitors.
But, it is possible it is introducing some weird behavior.

Despite all my tries, grid lines are being always shown only on primary monitor.
From other side, you had version where grids from both monitors were shown. May be, it is just offset for secondary monitor is being calculated incorrectly?

**Comment by Karolis-Sakavicius** [comment]
I've got a suspicion that work area adds the whole Y offset of the monitor + whatever UI elements are there here https://github.com/gTile/gTile/pull/262/files#diff-84dd6a83459413d5bb638cd885f20ef0dbebf2ebf10d0f35d1c9ca424b0ca06aR628, hence drawing the lines much lower than they should be.

Will get back later on or tomorrow, I'm not currently near my PC. But if you're curious, you could try to remove `topOffset` there and see what happens (it will start drawing on top of the GNOME UI elements if you have any, but that will prove the hypothesis). Thanks a lot for your assistance!

**Comment by Karolis-Sakavicius** [comment]
Ok, this should fix it. It relies completely on work areas, which should be calculated correctly, rather than on monitor offsets.

My second monitor was set up to have 0 Y offset, but yours was at the bottom, hence why I didn't see the issue on my part. Sorry for the back&forth here.

**Comment by scherepanov** [comment]
Yes, last commit fixed it.
I checked Preferences dialog, and it still opens with smallish size.  Have you committed change to make preferences dialog bigger?
Apart from this, PR is ready to be merged. Let me know if you are planning to do more in this PR.

**Comment by Karolis-Sakavicius** [comment]
I added `set_size_request` there and it enlarged the window for me
Screenshot from 2022-01-04 01-23-59
Screenshot from 2022-01-04 01-31-12

I'll change it to properties and let's see if it works then. 

Though I don't really like this, as this sets minimum size in pixels and this size can be different on different setups (languages, scaling, fonts, etc). I tried to fiddle with `expand` property and `vexpand-set`/`hexpand-set` which, as far as I understood from the docs, should scale the window to its contents. But it didn't work for me https://gjs-docs.gnome.org/gtk30/gtk.widget#property-expand, maybe because of the scrolled window.
One downside of the fixed approach is that this sets the minimum size of the window and it can't be shrunk manually.

**Comment by scherepanov** [comment]
Works fine!
Merging..
Thanks for contribution!

**Comment by scherepanov** [comment]
Are you interested in doing more work for gTile?
Based on my observation of user request, good feature to add to gTile would be saving current settings (including shortcuts) to a file, and restore settings from other file.
Current settings in gTile is more like an example that I put together to show what gTile can do. Idea was to let user to customize  presets and keyboard shortcuts.
It appears, vast majority of users are using default presets and settings. Configuring your own set of presets and keyboard shortcuts is not easy and time consuming.

What can be done:
Add tab to preferences for saving and restoring settings to a named file
Each setting will have a description, where user can put some info what keyboard shortcuts are defined and what presets they are linked to. Description is saved to a file along with other settings.
gTile will come with several files, each with it's own settings. User can browse files in drop-down list, and see description of what is in file. If user likes, he can press button load, and settings will be saved.
Similarly, after making changes to settings, user can choose name of file and save settings.
We will encourage people to submit their settings to repo. I am looking to include few files with settings into distribution, to allow wide audience to select what would fit best for their daily routing.
Of course, there are several other things that can be highly used by gTile user community.
Let me know if you are interested.

**Comment by Karolis-Sakavicius** [comment]
@scherepanov sure, I'm interested! And sounds like a very useful feature. Though my free time is quite limited now, so I can't make any promises, unfortunately. :( But I'll definitely contribute more. :)

**Pull request merged**